libkbfs/tlf_handle_resolve.go

```diff
@@ -0,0 +1,560 @@
+		other.ResolvesTo(ctx, codec, resolver, h)
+	if err != nil {
+		return err
+	}
+
+	handlePath := h.GetCanonicalPath()
+	otherPath := other.GetCanonicalPath()
+	if !handleResolvesToOther && !otherResolvesToHandle {
+		return MDMismatchError{
+			rev, h.GetCanonicalPath(), tlfID,
+			fmt.Errorf(
+				"MD contained unexpected handle path %s (%s -> %s) (%s -> %s)",
+				otherPath,
+				h.GetCanonicalPath(),
+				partialResolvedHandle.GetCanonicalPath(),
+				other.GetCanonicalPath(),
+				partialResolvedOther.GetCanonicalPath()),
+		}
+	}
+
+	if handlePath != otherPath {
+		log.CDebugf(ctx, "handle for %s resolved to %s",
+			handlePath, otherPath)
+	}
+	return nil
+}
+
+type resolvableAssertionWithChangeReport struct {
+	resolvableAssertion
+	changed chan struct{}
+}
+
+func (ra resolvableAssertionWithChangeReport) resolve(ctx context.Context) (
+	nameUIDPair, keybase1.SocialAssertion, error) {
+	nuid, sa, err := ra.resolvableAssertion.resolve(ctx)
+	if err != nil {
+		return nuid, sa, err
+	}
+	if nuid.name.String() != "" {
+		if nuid.name.String() != ra.assertion {
+			sendIfPossible(ra.changed)
+		}
+	} else if sa != (keybase1.SocialAssertion{}) {
+		if sa.String() != ra.assertion {
+			sendIfPossible(ra.changed)
+		}
+	}
+	return nuid, sa, nil
+}
+
+func sendIfPossible(ch chan struct{}) {
+	select {
+	case ch <- struct{}{}:
+	default:
+	}
+}
+
+type resolvableAssertion struct {
+	resolver   resolver
+	assertion  string
+	mustBeUser keybase1.UID
+}
+
+func (ra resolvableAssertion) resolve(ctx context.Context) (
+	nameUIDPair, keybase1.SocialAssertion, error) {
+	if ra.assertion == PublicUIDName {
+		return nameUIDPair{}, keybase1.SocialAssertion{}, fmt.Errorf("Invalid name %s", ra.assertion)
+	}
+	name, uid, err := ra.resolver.Resolve(ctx, ra.assertion)
+	if err == nil && ra.mustBeUser != keybase1.UID("") && ra.mustBeUser != uid {
+		// Force an unresolved assertion sinced the forced user doesn't match
+		err = NoSuchUserError{ra.assertion}
+	}
+	switch err := err.(type) {
+	default:
+		return nameUIDPair{}, keybase1.SocialAssertion{}, err
+	case nil:
+		return nameUIDPair{
+			name: name,
+			uid:  uid,
+		}, keybase1.SocialAssertion{}, nil
+	case NoSuchUserError:
+		socialAssertion, ok := externals.NormalizeSocialAssertion(ra.assertion)
+		if !ok {
+			return nameUIDPair{}, keybase1.SocialAssertion{}, err
+		}
+		return nameUIDPair{}, socialAssertion, nil
+	}
+}
+
+// ParseTlfHandle parses a TlfHandle from an encoded string. See
+// TlfHandle.GetCanonicalName() for the opposite direction.
+//
+// Some errors that may be returned and can be specially handled:
+//
+// TlfNameNotCanonical: Returned when the given name is not canonical
+// -- another name to try (which itself may not be canonical) is in
+// the error. Usually, you want to treat this as a symlink to the name
+// to try.
+//
+// NoSuchNameError: Returned when public is set and the given folder
+// has no public folder.
+func ParseTlfHandle(
```

Comment:
This function is called in `keybase_service_daemon`, and the `TlfNameNotCanonical` error type is explicitly checked.  So I think you need a version of this function that returns `TlfNameNotCanonical` when the alphabetical ordering is wrong.  Maybe that should be `ParseTlfHandle`, and this one should be called `ParseTlfHandleNoOrderChecking` or something like that.